Add comprehensive testscript for gh ext commandset

andyfeller opened the pull request:
Rather than multiple separate testscripts per `gh ext` subcommand, this commit adds a comprehensive test around `gh extension` commandset.

```shell
set -o pipefail && GH_ACCEPTANCE_TOKEN=$(gh auth token --hostname staffship-01.ghe.com) GH_ACCEPTANCE_HOST=staffship-01.ghe.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -run ^TestExtensions$ github.com/cli/cli/v2/acceptance  | tparse --all go test
```

resulting in:

```
┌───────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │           TEST            │             PACKAGE               │
│─────────┼─────────┼───────────────────────────┼───────────────────────────────────│
│  PASS   │   24.19 │ TestExtensions/commandset │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestExtensions            │ github.com/cli/cli/v2/acceptance  │
└───────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │ 24.72s  │ github.com/cli/cli/v2/acceptance │  --   │  2   │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

andyfeller committed:
Add comprehensive testscript for gh ext commandset

Rather than multiple separate testscripts per `gh ext` subcommand, this commit adds a comprehensive test around `gh extension` commandset.

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

williammartin reviewed (approved):
Strangely the first time I ran this I got:

```
     # Setup environment variables used for testscript (0.000s)
        > env EXT_NAME=${SCRIPT_NAME}-${RANDOM_STRING}
        > env EXT_SCRIPT=gh-${EXT_NAME}
        > env REPO=gh-${EXT_NAME}
        # Use gh as a credential helper (0.678s)
        > exec gh auth setup-git
        # Create local repository for extension (5.230s)
        > exec gh extension create $EXT_NAME
        > cd $REPO
        $WORK/gh-commandset-ClxBFjAnYI
        # Setup v1 executable baseline for extension (2.165s)
        > mv ../v1.sh $EXT_SCRIPT
        > chmod 777 $EXT_SCRIPT
        > exec git add $EXT_SCRIPT
        > exec git commit -m 'Setup extension as v1'
        [stderr]
        Author identity unknown

        *** Please tell me who you are.

        Run

          git config --global user.email "you@example.com"
          git config --global user.name "Your Name"

        to set your account's default identity.
        Omit --global to set the identity only in this repository.

        fatal: unable to auto-detect email address (got 'williammartin@William-GitHub-Laptop.(none)')
        [exit status 128]
        FAIL: testdata/extension/commandset.txtar:17: unexpected command failure
```

But afterwards:

```
➜  cli git:(andyfeller/testscript-extension) set -o pipefail && GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -coverprofile=coverage.out -coverpkg=./... -run ^TestExtensions$ github.com/cli/cli/v2/acceptance | tparse --all go test
┌───────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │           TEST            │             PACKAGE               │
│─────────┼─────────┼───────────────────────────┼───────────────────────────────────│
│  PASS   │   21.40 │ TestExtensions/commandset │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestExtensions            │ github.com/cli/cli/v2/acceptance  │
└───────────────────────────────────────────────────────────────────────────────────┘
```

Something to keep an eye on.

This looks really, really good though. One minor comment but I also don't think it **needs** to be addressed right now.

williammartin commented on acceptance/testdata/extension/commandset.txtar:
```diff
@@ -0,0 +1,66 @@
+# Setup environment variables used for testscript
```
Since this is relying on bash being available, maybe we would want to:

```
[!exec:bash] skip
```

Thinking about Windows users that aren't on WSL. I'm happy to defer and have a whole other piece of work that is specific to making these tests cross-platform though, if we ever cross that bridge.

Example here: https://github.com/rogpeppe/go-internal/blob/master/testscript/testdata/cond.txt

andyfeller replied:
Oof, good call. 🤔 There are Windows environments where Bash is available such as Git Bash, which should work.  That said, your recommendation is sound and will add that! ✨

andyfeller replied:
Going to leave some thoughts for the future on this:

- For gh ext testscript, since the only part where bash is actually required is for the part of running the extension.
- That portion could be refactored but obviously takes time and we're still clearing through things.
- Outside of refactoring that portion of the test out with the conditional, the only way is really switching over to creating and building a Go-based extension which is a bit more work to refactor.

andyfeller commented:
> Strangely the first time I ran this I got:
> 
> ```
>      # Setup environment variables used for testscript (0.000s)
>         > env EXT_NAME=${SCRIPT_NAME}-${RANDOM_STRING}
>         > env EXT_SCRIPT=gh-${EXT_NAME}
>         > env REPO=gh-${EXT_NAME}
>         # Use gh as a credential helper (0.678s)
>         > exec gh auth setup-git
>         # Create local repository for extension (5.230s)
>         > exec gh extension create $EXT_NAME
>         > cd $REPO
>         $WORK/gh-commandset-ClxBFjAnYI
>         # Setup v1 executable baseline for extension (2.165s)
>         > mv ../v1.sh $EXT_SCRIPT
>         > chmod 777 $EXT_SCRIPT
>         > exec git add $EXT_SCRIPT
>         > exec git commit -m 'Setup extension as v1'
>         [stderr]
>         Author identity unknown
> 
>         *** Please tell me who you are.
> 
>         Run
> 
>           git config --global user.email "you@example.com"
>           git config --global user.name "Your Name"
> 
>         to set your account's default identity.
>         Omit --global to set the identity only in this repository.
> 
>         fatal: unable to auto-detect email address (got 'williammartin@William-GitHub-Laptop.(none)')
>         [exit status 128]
>         FAIL: testdata/extension/commandset.txtar:17: unexpected command failure
> ```
> 
> But afterwards:
> 
> ```
> ➜  cli git:(andyfeller/testscript-extension) set -o pipefail && GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -coverprofile=coverage.out -coverpkg=./... -run ^TestExtensions$ github.com/cli/cli/v2/acceptance | tparse --all go test
> ┌───────────────────────────────────────────────────────────────────────────────────┐
> │  STATUS │ ELAPSED │           TEST            │             PACKAGE               │
> │─────────┼─────────┼───────────────────────────┼───────────────────────────────────│
> │  PASS   │   21.40 │ TestExtensions/commandset │ github.com/cli/cli/v2/acceptance  │
> │  PASS   │    0.00 │ TestExtensions            │ github.com/cli/cli/v2/acceptance  │
> └───────────────────────────────────────────────────────────────────────────────────┘
> ```
> 
> Something to keep an eye on.
> 
> This looks really, really good though. One minor comment but I also don't think it **needs** to be addressed right now.

Yes!  I was ruminating on the challenge writing these tests and `gh` design to defer to the user's `git` configuration for controlling behavior.  Outside of arranging some kind of alternative static configuration for `git` to leverage, we might need some kind of base line pre-requisite setup for who or whatever runs them. 🤷

andyfeller committed:
Skip extension testscript if bash isn't available

As this testscript creates a Bash-based script extension, the testscript should be skipped if it isn't on the path and executable.

Ideally, we would refactor this test to isolate that portion of the tests OR switch to a Go-based extension that can be compiled and run everywhere.

andyfeller committed:
Rename similar to other coarse commandset tests

Based on insights gained from reviewing conventions in #9815 with @jtmcg, I'm renaming this testscript to keep consistent with `gpg-key`, `label`, `ssh-key`, etc.

williammartin commented:
> Yes! I was ruminating on the challenge writing these tests and gh design to defer to the user's git configuration for controlling behavior. 

The weird thing is that I expect the baseline to be almost nothing, as if git had never been configured, since we set the `HOME` variable. I suppose there might be some system config like `/etc/gitconfig` though. But in any case I'm surprised that running the test back to back produced different results, especially since it doesn't seem to be doing anything different than other tests using `git`.

williammartin commented:
@andyfeller I rebased this on trunk and force pushed to fix the merge conflict. I figured it probably wouldn't cause you any issues since the branch will be merged. Hope that's ok!
